Lombiq.Tests.UI/Helpers/CloudflareHelper.cs

```diff
@@ -111,35 +125,37 @@ await ReliabilityHelper.DoWithRetriesAndCatchesAsync(
         finally
         {
             testOutputHelper.WriteLineTimestampedAndDebug(
-                "Current Cloudflare IP Access Rule reference count after the test (including this test): {0}.", _referenceCount);
+                "Current Cloudflare IP Access Rule reference count for IP {0} after the test (including this test): {1}.",
+                currentIp,
+                _referenceCounts[currentIp]);
 
             // Clean up the IP access rule.
-            if (_ipAccessRuleId != null && Interlocked.Decrement(ref _referenceCount) == 0)
+            if (_ipAccessRuleIds.TryGetValue(currentIp, out string value) && _referenceCounts.AddOrUpdate(currentIp, 0, (_, count) => count - 1) == 0)
             {
                 testOutputHelper.WriteLineTimestampedAndDebug(
                     "Removing the Cloudflare IP Access Rule for the IP {0} (Rule ID: {1}) since this test has the last reference to it.",
-                    _currentIp,
-                    _ipAccessRuleId);
+                    currentIp,
+                    value);
 
-                var oldIpAccessRuleId = _ipAccessRuleId;
+                var oldIpAccessRuleId = value;
 
-                var deleteSucceededResult = await DeleteIpAccessRuleWithRetriesAsync(cloudflareAccountId, _ipAccessRuleId);
+                var deleteSucceededResult = await DeleteIpAccessRuleWithRetriesAsync(cloudflareAccountId, value);
 
-                if (deleteSucceededResult.IsSuccess) _ipAccessRuleId = null;
+                if (deleteSucceededResult.IsSuccess) _ipAccessRuleIds.TryRemove(currentIp, out _);
 
-                ThrowIfNotSuccess(deleteSucceededResult, _currentIp, "couldn't be deleted");
+                ThrowIfNotSuccess(deleteSucceededResult, currentIp, "couldn't be deleted");
 
                 testOutputHelper.WriteLineTimestampedAndDebug(
                     "Removed the Cloudflare IP Access Rule for the IP {0} (Rule ID: {1}) since this test had the last reference to it.",
-                    _currentIp,
+                    currentIp,
                     oldIpAccessRuleId);
```

Comment:
Is there a point in using `oldIpAccessRuleId` in one place but `value` in another? Since here `oldIpAccessRuleId` is made equal to `value`.